bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,495 @@
+                try (Socket socket = getConnection(); OutputStream out = socket.getOutputStream()) {
+                    byte[] sendBuffer = command.getBytes(StandardCharsets.US_ASCII);
+                    // send the command
+                    out.write(sendBuffer, 0, sendBuffer.length);
+                }
+                // give time for mylink to process
+                Thread.sleep(CONNECTION_DELAY);
+            } catch (SocketTimeoutException e) {
+                logger.warn("Timeout sending command to mylink: {} Message: {}", command, e.getMessage());
+                future.completeExceptionally(new SomfyMyLinkException("Timeout sending command to mylink", e));
+            } catch (IOException e) {
+                logger.warn("Problem sending command to mylink: {} Message: {}", command, e.getMessage());
+                future.completeExceptionally(new SomfyMyLinkException("Problem sending command to mylink", e));
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting after sending command to mylink: {} Message: {}", command,
+                        e.getMessage());
+            }
+            future.complete(null);
+        });
+        return future;
+    }
+
+    private <T extends SomfyMyLinkResponseBase> CompletableFuture<@Nullable T> sendCommandWithResponse(String command, Class<T> responseType)
+            throws SomfyMyLinkException {
+        CompletableFuture<@Nullable T> future = new CompletableFuture<>();
+
+        commandExecutor.submit(() -> {
+            try {
+                logger.debug("Sending: {}", command);
+                try (Socket socket = getConnection();
+                        OutputStream out = socket.getOutputStream();
+                        BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.US_ASCII))) {
+                    byte[] sendBuffer = command.getBytes(StandardCharsets.US_ASCII);
+
+                    // send the command
+                    out.write(sendBuffer);
+                    out.flush();
+
+                    // now read the reply
+                    long startTime = System.currentTimeMillis();
+                    long elapsedTime = 0L;
+                    char[] readBuff = new char[1024];
+                    int readCount;
+                    String message = "";
+
+                    // while we are getting data ...
+                    while (((readCount = in.read(readBuff)) != -1)) {
+                        message += new String(readBuff, 0, readCount);
+
+                        // try and parse the message
+                        logger.debug("Got response: {}", message);
+
+                        try {
+                            T response = parseResponse(message, responseType);
+                            logger.debug("Got full message: {}", message);
+                            future.complete(response);
```

Comment:
A CompletableFuture can only be completed once, so the first time you complete it all further attempts to complete it will be ignored. So the timeout exception you attempt to complete below would not work if this statement completed successfully.

Also, if you got a successful response shouldn't you return here?

Reply:
i see. i think if it returns after the future.complete statement this will take care of the case when it gets a messages, completes and then returns it will never hit the timeout check in the loop.  sorry ... new to futures (and lots of java for that matter)

Reply:
updated. would appreciate a review however, as i'm still not 100% sure.